File: Examples/JavaScript/GameState/game_state.js
const fs = require('fs').promises;
const { execSync } = require('child_process');
const { createCanvas } = require('canvas');
const path = require('path');

class FlexonHelper {
    static runCommand(args) {
        try {
            return execSync(`flexon-cli ${args}`, { encoding: 'utf8' });
        } catch (error) {
            console.error(`Error running Flexon command: ${error.stderr}`);
            throw error;
        }
    }
}

class Vector2 {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}

class GameState {
    constructor() {
        this.version = "1.0.0";
        this.save_time = new Date().toISOString();
        this.player = {
            name: "DragonSlayer",
            level: 35,
            health: 100,
            mana: 85,
            position: new Vector2(123.45, 678.90),
            inventory: [
                { id: "sword_123", name: "Excalibur", damage: 50 },
                { id: "shield_456", name: "Dragon Shield", defense: 30 },
                { id: "potion_789", name: "Health Potion", healing: 100 }
            ]
        };
        this.world = {
            name: "Mythical Realm",
            current_zone: "Dragon's Lair",
            discovered_locations: [
                "Starting Village",
                "Dark Forest",
                "Crystal Cave",
                "Dragon's Lair"
            ],
            weather: "stormy",
            time_of_day: "night"
        };
        this.quests = {
            active: [
                { id: "quest_123", name: "Slay the Dragon", progress: 0.75 },
                { id: "quest_456", name: "Find the Sacred Sword", progress: 0.33 }
            ],
            completed: [
                "Rescue the Villagers",
                "Defeat the Dark Knight",
                "Find the Ancient Map"
            ]
        };
        this.achievements = {
            FirstBattle: true,
            DragonSlayer: false,
            LegendaryWeapon: true,
            ExplorerSupreme: false
        };
    }

    toJSON() {
        return {
            version: this.version,
            save_time: this.save_time,
            player: this.player,
            world: this.world,
            quests: this.quests,
            achievements: this.achievements
        };
    }
}

async function createGameScreenshot() {
    const width = 800;
    const height = 600;
    const canvas = createCanvas(width, height);
    const ctx = canvas.getContext('2d');

    // Draw background
    ctx.fillStyle = 'navy';
    ctx.fillRect(0, 0, width, height);

    // Draw border
    ctx.strokeStyle = 'white';
    ctx.lineWidth = 2;
    ctx.strokeRect(50, 50, width - 100, height - 100);

    // Draw text
    ctx.fillStyle = 'white';
    ctx.font = '30px Arial';
    ctx.textAlign = 'center';
    ctx.fillText("Dragon's Lair", width / 2, height / 2);

    // Draw player position
    ctx.beginPath();
    ctx.arc(width / 2, height / 2 + 50, 10, 0, Math.PI * 2);
    ctx.fillStyle = 'red';
    ctx.fill();

    // Save to file
    const buffer = canvas.toBuffer('image/png');
    await fs.writeFile('screenshot.png', buffer);
    return 'screenshot.png';
}

async function main() {
    console.log("Flexon Game State Example (JavaScript)");
    console.log("=====================================\n");

    // Create game state
    const gameState = new GameState();
    const stateFile = "gamestate.json";
    const saveFile = "savegame.flexon";

    // Create screenshot
    console.log("1. Creating game screenshot...");
    const screenshotFile = await createGameScreenshot();

    // Save game state
    console.log("2. Saving game state...");
    await fs.writeFile(stateFile, JSON.stringify(gameState.toJSON(), null, 2));

    // Create save package with state and screenshot
    console.log("3. Creating save package...");
    FlexonHelper.runCommand(`serialize -i ${stateFile} -i ${screenshotFile} -o ${saveFile} -e savegamekey`);

    // Load save package
    console.log("4. Loading save package...");
    await fs.mkdir("loaded_save", { recursive: true });
    FlexonHelper.runCommand(`deserialize -i ${saveFile} -o loaded_save -e savegamekey`);

    // Verify loaded state
    console.log("\nVerifying save data:");
    const loadedState = JSON.parse(
        await fs.readFile(path.join("loaded_save", "gamestate.json"), 'utf8')
    );

    console.log(`Player name: ${loadedState.player.name}`);
    console.log(`Level: ${loadedState.player.level}`);
    console.log(`Current zone: ${loadedState.world.current_zone}`);
    console.log(`Active quests: ${loadedState.quests.active.length}`);
    console.log(`Screenshot recovered: ${await fs.stat(path.join("loaded_save", "screenshot.png"))
        .then(() => true)
        .catch(() => false)}`);

    // Test different encryption algorithms
    console.log("\n5. Testing different encryption algorithms...");
    
    // ChaCha20
    console.log("\nTesting ChaCha20...");
    FlexonHelper.runCommand(`serialize -i ${stateFile} -o savegame_chacha20.flexon -e savegamekey ChaCha20`);
    
    // TripleDES
    console.log("Testing TripleDES...");
    FlexonHelper.runCommand(`serialize -i ${stateFile} -o savegame_tripledes.flexon -e savegamekey TripleDES`);

    // Compare file sizes
    const stats = await Promise.all([
        fs.stat(stateFile),
        fs.stat(saveFile),
        fs.stat('savegame_chacha20.flexon'),
        fs.stat('savegame_tripledes.flexon')
    ]);

    console.log("\nFile size comparison:");
    console.log(`Original JSON: ${stats[0].size} bytes`);
    console.log(`AES-256: ${stats[1].size} bytes`);
    console.log(`ChaCha20: ${stats[2].size} bytes`);
    console.log(`TripleDES: ${stats[3].size} bytes`);
}

main().catch(console.error);

File: Examples/Python/GameState/game_state.py
import json
import subprocess
import os
from datetime import datetime
from PIL import Image, ImageDraw, ImageFont
import numpy as np

class FlexonHelper:
    @staticmethod
    def run_command(args):
        try:
            result = subprocess.run(['flexon-cli'] + args.split(), 
                                  capture_output=True, text=True, check=True)
            return result.stdout
        except subprocess.CalledProcessError as e:
            print(f"Error running Flexon command: {e.stderr}")
            raise

class Vector2:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def to_dict(self):
        return {"x": self.x, "y": self.y}

class GameState:
    def __init__(self):
        self.version = "1.0.0"
        self.save_time = datetime.utcnow().isoformat()
        self.player = {
            "name": "DragonSlayer",
            "level": 35,
            "health": 100,
            "mana": 85,
            "position": Vector2(123.45, 678.90).to_dict(),
            "inventory": [
                {"id": "sword_123", "name": "Excalibur", "damage": 50},
                {"id": "shield_456", "name": "Dragon Shield", "defense": 30},
                {"id": "potion_789", "name": "Health Potion", "healing": 100}
            ]
        }
        self.world = {
            "name": "Mythical Realm",
            "current_zone": "Dragon's Lair",
            "discovered_locations": [
                "Starting Village",
                "Dark Forest",
                "Crystal Cave",
                "Dragon's Lair"
            ],
            "weather": "stormy",
            "time_of_day": "night"
        }
        self.quests = {
            "active": [
                {"id": "quest_123", "name": "Slay the Dragon", "progress": 0.75},
                {"id": "quest_456", "name": "Find the Sacred Sword", "progress": 0.33}
            ],
            "completed": [
                "Rescue the Villagers",
                "Defeat the Dark Knight",
                "Find the Ancient Map"
            ]
        }
        self.achievements = {
            "FirstBattle": True,
            "DragonSlayer": False,
            "LegendaryWeapon": True,
            "ExplorerSupreme": False
        }

    def to_dict(self):
        return {
            "version": self.version,
            "save_time": self.save_time,
            "player": self.player,
            "world": self.world,
            "quests": self.quests,
            "achievements": self.achievements
        }

def create_game_screenshot():
    # Create a simple game screenshot
    width, height = 800, 600
    image = Image.new('RGB', (width, height), color='navy')
    draw = ImageDraw.Draw(image)
    
    # Draw some game elements
    draw.rectangle([50, 50, 750, 550], outline='white')
    draw.text((400, 300), "Dragon's Lair", fill='white', anchor='mm')
    
    # Draw player position
    draw.ellipse([390, 290, 410, 310], fill='red')
    
    # Save the image
    image.save('screenshot.png')
    return 'screenshot.png'

def main():
    print("Flexon Game State Example (Python)")
    print("==================================\n")

    # Create game state
    game_state = GameState()
    state_file = "gamestate.json"
    save_file = "savegame.flexon"

    # Create screenshot
    print("1. Creating game screenshot...")
    screenshot_file = create_game_screenshot()

    # Save game state
    print("2. Saving game state...")
    with open(state_file, 'w') as f:
        json.dump(game_state.to_dict(), f, indent=2)

    # Create save package with state and screenshot
    print("3. Creating save package...")
    FlexonHelper.run_command(f"serialize -i {state_file} -i {screenshot_file} -o {save_file} -e savegamekey")

    # Load save package
    print("4. Loading save package...")
    os.makedirs("loaded_save", exist_ok=True)
    FlexonHelper.run_command(f"deserialize -i {save_file} -o loaded_save -e savegamekey")

    # Verify loaded state
    print("\nVerifying save data:")
    with open(os.path.join("loaded_save", "gamestate.json"), 'r') as f:
        loaded_state = json.load(f)

    print(f"Player name: {loaded_state['player']['name']}")
    print(f"Level: {loaded_state['player']['level']}")
    print(f"Current zone: {loaded_state['world']['current_zone']}")
    print(f"Active quests: {len(loaded_state['quests']['active'])}")
    print(f"Screenshot recovered: {os.path.exists(os.path.join('loaded_save', 'screenshot.png'))}")

    # Test different encryption algorithms
    print("\n5. Testing different encryption algorithms...")
    
    # ChaCha20
    print("\nTesting ChaCha20...")
    FlexonHelper.run_command(f"serialize -i {state_file} -o savegame_chacha20.flexon -e savegamekey ChaCha20")
    
    # TripleDES
    print("Testing TripleDES...")
    FlexonHelper.run_command(f"serialize -i {state_file} -o savegame_tripledes.flexon -e savegamekey TripleDES")

    # Compare file sizes
    print("\nFile size comparison:")
    print(f"Original JSON: {os.path.getsize(state_file)} bytes")
    print(f"AES-256: {os.path.getsize(save_file)} bytes")
    print(f"ChaCha20: {os.path.getsize('savegame_chacha20.flexon')} bytes")
    print(f"TripleDES: {os.path.getsize('savegame_tripledes.flexon')} bytes")

if __name__ == "__main__":
    main()
